enhancements/authentication/service-ca-cert-generation-for-statefulset-pods.md

```diff
@@ -0,0 +1,327 @@
+---
+title: service-ca-cert-generation-for-statefulset-pods
+authors:
+  - "@mtrmac"
+reviewers:
+  - TBD
+approvers:
+  - TBD
+creation-date: 2021-03-01
+last-updated: 2021-03-01
+status: implementable
+see-also:
+replaces:
+superseded-by:
+---
+
+# Service CA Certificate Generation for StatefulSet Pods
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Operational readiness criteria is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+The Service CA-generated certificates _for headless services_ include
+`*.${service.name}.${service.namespace}.svc` and
+`*.${service.name}.${service.namespace}.svc.cluster.local`
+as subjects,
+to allow TLS-protected communication between StatefulSet pods.
+
+## Motivation
+
+StatefulSet-managed pods often need to communicate with each other.
+This traffic may need to be TLS-protected; because it is cluster-internal,
+and necessary for deploying such a StatefulSet, letting the cluster automatically
+manage the certificates is a natural extension to the same feature already
+provided by the Service CA to allow TLS use with services.
+
+### Goals
+
+Allow deploying StatefulSets with pods that communicate with each other using TLS,
+without having to manually generate certificates for these pods.
+
+### Non-Goals
+
+- Provide certificates for `127.0.0.1`, `localhost`, or NodeIP service endpoints
+
+  (See discussion in “Implementation Details/Notes/Constraints”)
+
+- Provide individual certificates that uniquely identify individual pods in the StatefulSet,
+  protecting the intra-StatefulSet communication against re-routing to unintended members
+  of that StatefulSet
+  (or to a different StatefulSet attached to the same headless service).
+
+## Proposal
+
+Currently, a service annotated with
+a `service.beta.openshift.io/serving-cert-secret-name: $secretName`
+requests the `service-ca` controller to generate a secret `$secretName`
+containing a certificate+key pair for `${service.name}.${service.namespace}.svc`
+(and a `….cluster.local`) as subjects.
+
+For headless services, generate certificates that include two more subjects,
```

Comment:
Compatibility with existing users.

Headless services are, by naive clients, accessible using exactly the same ${service.name}.${service.namespace}.svc host names — the only difference is that for headless services that DNS entry points directly to the pods as multiple A/AAAA records, not to the service’s non-existent ClusterIP.

So, the pre-existing certificate generation can be used both for ClusterIP an headless services in the same way, and reading https://access.redhat.com/documentation/en-us/openshift_container_platform/4.7/html/security_and_compliance/configuring-certificates#add-service-certificate_service-serving-certificate does not suggest that headless services are not expected to work.

---

I could see an argument that it would be cleaner for a service to _either_ be used by the single service hostname (to contact any pod) _or_ to use the StatefulSet-generated individual hostnames for directed communication, but I’m not aware of any implementation restriction that would prevent a single service from being used for both at the same time; e.g. https://github.com/kubernetes/dns/blob/master/docs/specification.md#241---aaaaa-records says both DNS names must work simultaneously.

----

A possible alternative would be to let the user choose (generate a certificate for only the service host name, only the wildcard, or both); that would also mitigate the security risks of wildcards somewhat, in that we would not start generating certificates with completely unnecessary wildcard certificates. OTOH it’s again one more option for the user to deal with.